pisa/stages/flux/pi_mceq_barr.py

```diff
@@ -105,267 +186,339 @@ def __init__(self,
+
+
+    def antipion_production(self, barr_var, pion_ratio):
+        '''
+        Combine pi+ param and pi+/pi- ratio to get pi- param
+        '''
+        return ((1 + barr_var)/(1 + pion_ratio)) - 1
 
 
     @profile
     def compute_function(self):
 
         self.data.data_specs = self.calc_specs
 
-        barr_a = self.params.barr_a.value.m_as('dimensionless')
-        barr_b = self.params.barr_b.value.m_as('dimensionless')
-        barr_c = self.params.barr_c.value.m_as('dimensionless')
-        barr_d = self.params.barr_d.value.m_as('dimensionless')
-        barr_e = self.params.barr_e.value.m_as('dimensionless')
-        barr_f = self.params.barr_f.value.m_as('dimensionless')
-        barr_g = self.params.barr_g.value.m_as('dimensionless')
-        barr_h = self.params.barr_h.value.m_as('dimensionless')
-        barr_i = self.params.barr_i.value.m_as('dimensionless')
-        barr_w = self.params.barr_w.value.m_as('dimensionless')
-        barr_y = self.params.barr_y.value.m_as('dimensionless')
-        barr_z = self.params.barr_z.value.m_as('dimensionless')
+        #
+        # Get params
+        #
+
+        # Spectral index (and required energy pivot)
+        delta_index = self.params.delta_index.value.m_as("dimensionless")
+        energy_pivot = self.params.energy_pivot.value.m_as("GeV")
+
+        # pions
+        pion_ratio = self.params.pion_ratio.value.m_as('dimensionless')
+        barr_a_Pi = self.params.barr_a_Pi.value.m_as('dimensionless')
+        barr_b_Pi = self.params.barr_b_Pi.value.m_as('dimensionless')
+        barr_c_Pi = self.params.barr_c_Pi.value.m_as('dimensionless')
+        barr_d_Pi = self.params.barr_d_Pi.value.m_as('dimensionless')
+        barr_e_Pi = self.params.barr_e_Pi.value.m_as('dimensionless')
+        barr_f_Pi = self.params.barr_f_Pi.value.m_as('dimensionless')
+        barr_g_Pi = self.params.barr_g_Pi.value.m_as('dimensionless')
+        barr_h_Pi = self.params.barr_h_Pi.value.m_as('dimensionless')
+        barr_i_Pi = self.params.barr_i_Pi.value.m_as('dimensionless')
+        #TODO also have indepednent priors on antipion params?
+
+        #kaons
+        barr_w_K = self.params.barr_w_K.value.m_as('dimensionless')
+        barr_x_K = self.params.barr_x_K.value.m_as('dimensionless')
+        barr_y_K = self.params.barr_y_K.value.m_as('dimensionless')
+        barr_z_K = self.params.barr_z_K.value.m_as('dimensionless')
+        barr_w_antiK = self.params.barr_w_antiK.value.m_as('dimensionless')
+        barr_x_antiK = self.params.barr_x_antiK.value.m_as('dimensionless')
+        barr_y_antiK = self.params.barr_y_antiK.value.m_as('dimensionless')
+        barr_z_antiK = self.params.barr_z_antiK.value.m_as('dimensionless')
+
+        # Map the user parameters into the Barr +/- params
+        # pi- production rates is restricted by the pi-ratio, just as in arXiv:0611266
+        gradient_params_mapping = collections.OrderedDict()
+        gradient_params_mapping["a+"] = barr_a_Pi
```

Comment:
would be less error prone code is we did first directly `gradient_params_mapping["a+"]` = self.params.barr_a_Pi.value.m_as('dimensionless')

and the followed in a loop like `for name in self.barr_param_names: gradient_params_mapping[name + '-'] = self.antipion_production(gradient_params_mapping[name + '+'])` or something along these lines

Reply:
Done